### PR TITLE
APIM-162 first implementation of mqtt5 endpoint

### PR DESCRIPTION
**Pull request by guillaumelamirand**
## Issue

https://graviteecommunity.atlassian.net/browse/APIM-162

## Description

First implementation of mqtt5 endpoint.

## Additional context

🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.z6.web.core.windows.net/apim-162-push-mqtt-messages/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-sqywxsrsae.chromatic.com)

### CODE REVIEW

File: gravitee-apim-plugin/gravitee-apim-plugin-endpoint/gravitee-apim-plugin-endpoint-mqtt5/docker/Makefile [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	cd ./quick-setup/apim-hivemq && docker-compose start
+
+stop-all: ## Stop API Management with hivemq broker
+	cd ./quick-setup/apim-hivemq && docker-compose stop
+
+hivemq: ## Run only the hivemq service
+	cd ./quick-setup/apim-hivemq && docker-compose up --force-recreate -d hivemq
+
+start-hivemq: ## Start only the hivemq service
+	cd ./quick-setup/apim-hivemq && KAFKA_HOST=localhost docker-compose start hivemq
```
Comment (jhaeyaert) [inline gravitee-apim-plugin/gravitee-apim-plugin-endpoint/gravitee-apim-plugin-endpoint-mqtt5/docker/Makefile]: I think we can remove all reference to kafka 😁

File: gravitee-apim-plugin/gravitee-apim-plugin-endpoint/gravitee-apim-plugin-endpoint-mqtt5/src/main/resources/schema/schema-form.json [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                },
+                "retained": {
+                    "title": "retained",
+                    "description": "Define if the retain flag must be set to every publish messages. The broker stores the last retained message.",
+                    "type": "boolean",
+                    "default": false
+                },
+                "messageExpiryInterval": {
+                    "title": "messageExpiryInterval",
+                    "description": "This interval defines the period of time that the broker stores the PUBLISH message for any matching subscribers that are not currently connected. When no message expiry interval is set, the broker must store the message for matching subscribers indefinitely. When the retained=true option is set on the PUBLISH message, this interval also defines how long a message is retained on a topic.",
```
Comment (jhaeyaert) [inline gravitee-apim-plugin/gravitee-apim-plugin-endpoint/gravitee-apim-plugin-endpoint-mqtt5/src/main/resources/schema/schema-form.json:58]: Is it possible to indicate the time unit (I supposed it was in milliseconds but digging the code I sa it was seconds).
Also, I'm wondering if, for consistency, we should always use millisecond time unit and make the translation when required?

File: gravitee-apim-plugin/gravitee-apim-plugin-endpoint/gravitee-apim-plugin-endpoint-mqtt5/src/main/java/io/gravitee/plugin/endpoint/mqtt5/Mqtt5EndpointConnector.java [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+                                                                messageBuilder.metadata(
+                                                                    Maps
+                                                                        .<String, Object>builder()
+                                                                        .put("topic", mqttPublish.getTopic())
+                                                                        .put("type", mqttPublish.getType().name())
+                                                                        .put("qos", mqttPublish.getQos().getCode())
+                                                                        .put("retain", mqttPublish.isRetain())
+                                                                        .put("contentType", mqttPublish.getContentType())
+                                                                        .put(
+                                                                            "messageExpiryInterval",
```
Comment (jhaeyaert) [inline gravitee-apim-plugin/gravitee-apim-plugin-endpoint/gravitee-apim-plugin-endpoint-mqtt5/src/main/java/io/gravitee/plugin/endpoint/mqtt5/Mqtt5EndpointConnector.java:118]: I wasn't able to properly test the `messageExpiryInterval` feature.
Maybe it is too soon for this PR and we will have to spend more time on it later. From what I saw here (https://www.hivemq.com/blog/mqtt5-essentials-part4-session-and-message-expiry/), `messageExpiryInterval` can be used to implement a kind of retention mechanism. Depending whether the retain flag is set or not, the behavior is not the same. Also, it seems that other configuration are required to make it work properly (sessionExpiry, cleanStart, ...). They appear to be defined at connection level, ex:

```java
 rxClient                         
     .connect(MqttConnect.DEFAULT.extend().sessionExpiryInterval(100000).cleanStart(false).build())
```

Note: I didn't test it. Just for note the point for a later independent task.

**Review by jhaeyaert — APPROVED** [review]
All good for me regarding publish / consume of messages 👍 
I let you have a look at the notes I left (few typos) and thoughts that should be handled in other independent tasks.

File: gravitee-apim-plugin/gravitee-apim-plugin-endpoint/gravitee-apim-plugin-endpoint-mqtt5/docker/Makefile (hunk 1 above)
Comment (guillaumelamirand) [reply gravitee-apim-plugin/gravitee-apim-plugin-endpoint/gravitee-apim-plugin-endpoint-mqtt5/docker/Makefile]: of course...

**Pull request merged**